src/network_service/sending_targets_cache.rs

```diff
@@ -0,0 +1,132 @@
+
+const MAX_RESENDS: u8 = 3;
+
+enum TargetState {
+    Sending(u8),
+    Failed(u8),
+    Sent,
+}
+
+impl TargetState {
+    pub fn is_complete(&self) -> bool {
+        match *self {
+            TargetState::Failed(x) => x > MAX_RESENDS,
+            TargetState::Sent => true,
+            TargetState::Sending(_) => false,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct SendingTargetsCache {
+    cache: HashMap<u64, Vec<(ConnectionInfo, TargetState)>>,
+}
+
+impl SendingTargetsCache {
+    pub fn insert_message(
+        &mut self,
+        msg_id: u64,
+        initial_targets: Vec<ConnectionInfo>,
+        dg_size: usize,
+    ) {
+        let targets = initial_targets
+            .into_iter()
+            .enumerate()
+            .map(|(idx, tgt_info)| {
+                (
+                    tgt_info,
+                    if idx < dg_size {
+                        TargetState::Sending(0)
+                    } else {
+                        TargetState::Failed(0)
+                    },
+                )
+            })
+            .collect();
+        let _ = self.cache.insert(msg_id, targets);
+    }
+
+    fn target_states(&self, msg_id: u64) -> impl Iterator<Item = &(ConnectionInfo, TargetState)> {
+        self.cache.get(&msg_id).into_iter().flatten()
+    }
+
+    fn target_states_mut(
+        &mut self,
+        msg_id: u64,
+    ) -> impl Iterator<Item = &mut (ConnectionInfo, TargetState)> {
+        self.cache.get_mut(&msg_id).into_iter().flatten()
+    }
+
+    fn fail_target(&mut self, msg_id: u64, target: SocketAddr) {
+        let _ = self
+            .target_states_mut(msg_id)
+            .find(|(info, _state)| info.peer_addr() == target)
+            .map(|(_info, state)| match *state {
+                TargetState::Failed(_) => {
+                    log_or_panic!(LogLevel::Error, "Got a failure from a failed target!");
+                }
+                TargetState::Sending(x) => {
+                    *state = TargetState::Failed(x + 1);
+                }
+                TargetState::Sent => {
+                    log_or_panic!(
+                        LogLevel::Error,
+                        "A target that should no longer fail - failed!"
+                    );
+                }
+            });
+    }
+
+    fn take_next_target(&mut self, msg_id: u64) -> Option<ConnectionInfo> {
+        self.target_states_mut(msg_id)
+            .filter(|(_info, state)| !state.is_complete())
+            .filter_map(|(info, state)| match state {
+                TargetState::Failed(x) => Some((info, *x, state)),
+                _ => None,
+            })
+            .min_by_key(|(_info, num, _state)| *num)
+            .map(|(info, num, state)| {
+                *state = TargetState::Sending(num);
+                info
+            })
+            .cloned()
+    }
+
+    fn should_drop(&self, msg_id: u64) -> bool {
+        self.target_states(msg_id)
+            .all(|(_info, state)| state.is_complete())
```

Comment:
There may be a bug here.
I think we will not drop the message for a delivery group of size 3, if 3 out of 7 succeed, not all will be complete.
This is the valid should_drop to use in case of failure, however in case of success we should drop if no message is in `Sending` state.

Reply:
I guess if we move the should_drop after take_next_target in target_failed, we can use none(is_sending) as a should droup for both success and failure.

Reply:
Or rather, we should drop if `delivery_group_size` targets are in `Sent` state, regardless of the state of other targets. You're right it's a bug :+1:

Reply:
Well they are equivalent, we should always have delivery_group_size in sent or sending state.
so when none are sending they have all been sent or failed.

I'd rather avoid using delivery group size twice.
i had a local patch to see if it was impacting perforamnce, it did not:

```
             TargetState::Sending(_) => false,
         }
     }
+
+    pub fn is_sending(&self) -> bool {
+        match *self {
+            TargetState::Failed(_) |
+            TargetState::Sent => false,
+            TargetState::Sending(_) => true,
+        }
+    }
 }
 
 #[derive(Default)]
@@ -106,16 +114,18 @@
 
     fn should_drop(&self, msg_id: u64) -> bool {
         self.target_states(msg_id)
-            .all(|(_info, state)| state.is_complete())
+            .all(|(_info, state)| !state.is_sending())
     }
 
     pub fn target_failed(&mut self, msg_id: u64, target: SocketAddr) -> Option<ConnectionInfo> {
         self.fail_target(msg_id, target);
+
+        let next = self.take_next_target(msg_id);
         if self.should_drop(msg_id) {
             let _ = self.cache.remove(&msg_id);
         }
-        // if we dropped the msg_id above, this would have returned None even if we hadn't
-        self.take_next_target(msg_id)
+
+        next
     }
 ```

Reply:
They are equivalent currently, but we should make sure that we made this assumption. I'll apply your patch (looks good to me) and add a comment so that if we ever change the logic, we remember about it.

Reply:
There is another way of looking at it that avoid us having to make that direct assumption:
The cached information is only useful as long as we have a message with Sending because we may either receive a failure or success. Once no message are in Sending, this cache will never be accessed anymore.

Then what we do is try sending delivery_group_size, and it will complete either by succeeding to send to delivery_group_size, or by failing enough time trying.